server/backends/invocationdb/invocationdb.go

```diff
@@ -40,10 +45,42 @@ func getACL(i *tables.Invocation) *aclpb.ACL {
 	return perms.ToACLProto(&uidpb.UserId{Id: i.UserID}, i.GroupID, i.Perms)
 }
 
+func upsertInvocationForRegisterInvocationAttempt(d *db.DB, ti *tables.Invocation) (bool, error) {
+	updates, err := db.FillUpdatesForOnConflict(d, ti, map[string]any{
+		"attempt": gorm.Expr(
+			`CASE WHEN "Invocations".attempt = 0 THEN 2 ELSE "Invocations".attempt + 1 END`,
+		),
+	})
+	if err != nil {
+		return false, err
+	}
+	result := d.Clauses(
+		clause.OnConflict{
+			Columns: []clause.Column{{Name: clause.PrimaryKey}},
+			Where: clause.Where{
+				Exprs: []clause.Expression{
+					gorm.Expr(`"Invocations".invocation_status <> ?`, int64(inspb.InvocationStatus_COMPLETE_INVOCATION_STATUS)),
+					gorm.Expr(`"Invocations".updated_at_usec > ?`, d.NowFunc().Add(time.Hour*-4).UnixMicro()),
+				},
+			},
+			DoUpdates: updates,
+		},
+		clause.Returning{},
+	).Create(ti)
+	return result.RowsAffected > 0, result.Error
+}
```

Comment:
This code feels difficult to grok because it is using some of the more advanced GORM functionality (`clause.OnConflict{}`, `clause.Returning{}`, Session, etc), which itself maps to some advanced-ish SQL.

Would it work to do a simple transaction like this?

```
Transaction("upsert_invocation", func (tx) {
  // Get existing row; use SELECT ... FOR UPDATE construct
  // to lock the row during the transaction.
  // Include iid in WHERE clause, and maybe attempt number
  // if we know this is not the first attempt.
  existing := &Invocation{}
  err := tx.Raw(`SELECT * FROM Invocations WHERE ... ` +db.SelectForUpdateModifier()).Take(existing).Error

  // check err; but allow NotFound
  /* ... */
  
  // If we know the current attempt number, check to make sure
  // we weren't preempted by a more recent attempt
  if (ourAttempt > 0 && existing != nil && existing.Attempt > ourAttempt) {
    return // a newer attempt exists; do nothing.
  }
  if (!existing) {
    // Create empty invocation row for (iid, attempt)
    /* ... */
  }
  tx.Updates(invocation)
})
```

and then maybe wrap this with a retry+backoff in case the lock failed.

Reply:
Could I just add comments to make the clause stuff easier to understand? It doesn't feel simpler to me to add all of this application logic and a transaction and then have to add a retry and backoff as well. The clauses are pretty straightforward, imo, and I'm happy to add comments explaining them.

Reply:
I'd write the raw SQL for it but we *need* the ORM to ensure code maintainability as we add/remove columns from tables.